lib/ecto/query/builder/preload.ex

```diff
@@ -156,4 +157,126 @@ defmodule Ecto.Query.Builder.Preload do
   def apply(query, preloads, assocs) do
     apply(Ecto.Queryable.to_query(query), preloads, assocs)
   end
+
+  @doc """
+  Called at runtime to assemble preload.
+  """
+  def preload!(query, preload) do
+    {preloads, assocs} = expand(preload, query)
+    apply(query, Enum.reverse(preloads), Enum.reverse(assocs))
+  end
+
+  @doc """
+  Expands preloads at runtime.
+  """
+  def expand(preloads, query) do
+    expand(preloads, query, :both, [], [])
+  end
+
+  defp expand(atom, _query, _mode, preloads, assocs) when is_atom(atom) do
+    {[atom|preloads], assocs}
+  end
+
+  defp expand(list, query, mode, preloads, assocs) when is_list(list) do
+    Enum.reduce(list, {preloads, assocs}, fn item, acc ->
+      expand_each(item, query, mode, acc)
+    end)
+  end
+
+  defp expand(other, _query, _mode, _preloads, _assocs) do
+    raise ArgumentError, "`#{inspect(other)}` is not a valid preload expression."
+  end
```

Comment:
```suggestion
  def expand(preloads, query) when is_atom(preloads) or is_list(preloads) do
    expand_each(preloads, query, :both, [], [])
  end

  def expand(other, _query) do
    raise ArgumentError, "`#{inspect(other)}` is not a valid preload expression."
  end
```

Reply:
Just want to make sure I understand this suggestion correctly.

Right now, `expand_each` doesn't have a case that operates on a list -- it expects either an atom or a 2-element tuple. When `expand_each` needs to process a list, it calls back into `expand`. So basically, we need the `Enum.reduce(...)` clause currently handled by `defp expand(list, query, mode, preloads, assocs) when is_list(list)` somewhere.

My goal here was to emulate as much as possible the strategy that `escape` uses to process the AST. Moving the reduce into `expand_each` would, I think, mean that something like this would start working:

```elixir
# expressions interpolated at root are handled by `expand`
preload(^[:foo, [:bar, [:baz]])
```

whereas this wouldn't work:

```elixir
# expressions not interpolated at root are handled by `escape`
preload([:foo, [:bar, [:baz]])
```

Reply:
You are right, please ignore this suggestion then except for the change the error message. :)